src/componentprocessor/index.ts

```diff
@@ -34,12 +51,93 @@ export class ComponentDefinition {
     }
 }
 
-export async function getComponentDefinitions(rootDir: string) {
+export async function getComponentDefinitionMap(rootDir: string) {
     const componentsPattern = rootDir + "/components/**/*.xml";
     const xmlFiles: string[] = fg.sync(componentsPattern, {});
 
     let defs = xmlFiles.map(file => new ComponentDefinition(file));
     let parsedPromises = defs.map(async def => def.parse());
 
-    return pSettle(parsedPromises);
+    return processXmlTree(pSettle(parsedPromises));
+}
+
+async function processXmlTree(
+    settledPromises: Promise<pSettle.SettledResult<ComponentDefinition>[]>
+) {
+    let nodeDefs = await settledPromises;
+    let nodeDefMap = new Map<string, ComponentDefinition>();
+
+    // create map of just ComponentDefinition objects
+    nodeDefs.map(node => {
+        if (node.isFulfilled && !node.isRejected) {
+            nodeDefMap.set(node.value!.name!, node.value!);
+        }
+    });
+
+    // recursively create an inheritance stack for each component def and build up
+    // the component backwards from most extended component first
+    let inheritanceStack: ComponentDefinition[] = [];
+
+    nodeDefMap.forEach((node, nodeName) => {
+        if (node && node.processed === false) {
+            let xmlNode = node.xmlNode;
+            inheritanceStack.push(node);
+            //builds inheritance stack
+            while (xmlNode && xmlNode.attr.extends) {
+                let superNode = nodeDefMap.get(xmlNode.attr.extends);
+                if (superNode) {
+                    inheritanceStack.push(superNode);
+                    xmlNode = superNode.xmlNode;
+                } else {
+                    xmlNode = undefined;
+                }
+            }
+
+            let inheritedFields: ComponentField = {};
+            // pop the stack & build our component
+            // we can safely assume nodes are valid ComponentDefinition objects
+            while (inheritanceStack.length > 0) {
+                let newNode = inheritanceStack.pop();
+                if (newNode) {
+                    if (newNode.processed) {
+                        inheritedFields = newNode.fields;
+                    } else {
+                        let nodeFields = getFields(newNode.xmlNode!);
+                        // we will get run-time error if any fields are duplicated
+                        // between inherited components, but here we will retain
+                        // the original value without throwing an error for simplicity
+                        inheritedFields = { ...nodeFields, ...inheritedFields };
```

Comment:
Can you add a ToDo for the case when fields are duplicated in parent and children? We should throw an exception or something like that